packages/relay-compiler/core/RelayFindGraphQLTags.js

```diff
@@ -0,0 +1,125 @@
+} from '../language/RelayLanguagePluginInterface';
+
+export type GraphQLTagFinderOptions = {|
+  validateNames: boolean,
+|};
+
+const cache = new RelayCompilerCache('RelayFindGraphQLTags', 'v1');
+
+function memoizedFind(
+  tagFinder: GraphQLTagFinder,
+  text: string,
+  baseDir: string,
+  file: File,
+  options: GraphQLTagFinderOptions,
+): Array<string> {
+  invariant(
+    file.exists,
+    'RelayFindGraphQLTags: Called with non-existent file `%s`',
+    file.relPath,
+  );
+  return cache.getOrCompute(
+    file.hash + (options.validateNames ? '1' : '0'),
+    find.bind(null, tagFinder, text, path.join(baseDir, file.relPath), options),
+  );
+}
+
+function find(
+  tagFinder: GraphQLTagFinder,
+  text: string,
+  absPath: string,
+  {validateNames}: GraphQLTagFinderOptions,
+): Array<string> {
+  const tags = tagFinder(text, absPath);
+  if (validateNames) {
+    const moduleName = getModuleName(absPath);
+    tags.forEach(tag => validateTemplate(tag, moduleName, absPath));
+  }
+  return tags.map(tag => tag.template);
+}
+
+function validateTemplate(
+  {template, keyName, sourceLocationOffset}: GraphQLTag,
+  moduleName: string,
+  filePath: string,
+) {
+  const ast = graphql.parse(
+    new graphql.Source(template, filePath, sourceLocationOffset),
+  );
+  ast.definitions.forEach((def: any) => {
+    invariant(
+      def.name,
+      'RelayFindGraphQLTags: In module `%s`, a definition of kind `%s` requires a name.',
+      moduleName,
+      def.kind,
+    );
+    const definitionName = def.name.value;
+    if (def.kind === 'OperationDefinition') {
+      const operationNameParts = definitionName.match(
+        /^(.*)(Mutation|Query|Subscription)$/,
+      );
+      invariant(
+        operationNameParts && definitionName.startsWith(moduleName),
+        'RelayFindGraphQLTags: Operation names in graphql tags must be prefixed ' +
+          'with the module name and end in "Mutation", "Query", or ' +
+          '"Subscription". Got `%s` in module `%s`.',
+        definitionName,
+        moduleName,
+      );
+    } else if (def.kind === 'FragmentDefinition') {
+      if (keyName) {
+        invariant(
+          definitionName === moduleName + '_' + keyName,
+          'RelayFindGraphQLTags: Container fragment names must be ' +
+            '`<ModuleName>_<propName>`. Got `%s`, expected `%s`.',
+          definitionName,
+          moduleName + '_' + keyName,
+        );
+      } else {
+        invariant(
+          definitionName.startsWith(moduleName),
+          'RelayFindGraphQLTags: Fragment names in graphql tags must be prefixed ' +
+            'with the module name. Got `%s` in module `%s`.',
+          definitionName,
+          moduleName,
+        );
+      }
+    }
+  });
+}
+
+// TODO: Not sure why this is defined here rather than imported, is it so that it doesn’t get stripped in prod?
+function invariant(condition, msg, ...args) {
+  if (!condition) {
+    throw new Error(util.format(msg, ...args));
+  }
+}
```

Comment:
Any idea why this needed to be defined in `FindGraphQLTags.js` rather than importing it?

Reply:
I think it was to allow the compiler to be more restrictive in its dependencies, but I'm not sure